### PR TITLE
Registering VisualVM-ready AppearanceController

### PR DESCRIPTION
**Pull request by JaroslavTulach**
Providing specific `AppearanceController` implementation to adjust the popup menu & co. to needs of VisualVM. Requires update to `lib.profiler.ui` version `1.147` or newer.

### CODE REVIEW

**Comment by thurka** [comment]
You need to also provide new version of [NetBeans platform and profiler binaries](https://github.com/oracle/visualvm/blob/master/visualvm/nb90_visualvm_11122017.zip), otherwise VisualVM cannot be build after this merge request is merged.

File: visualvm/build-nb-ml.sh [hunk 1]
```diff
@@ -1,13 +1,76 @@
-#!/bin/sh
+#!/bin/bash
+
+REV=41da26b7de7506c8
+REV_L10N=1180ea1ceb30
+ZIPNAME=nb90_visualvm_`date "+%d%m%Y"`
```
Comment (JaroslavTulach) [inline visualvm/build-nb-ml.sh]: This is a build script that prepares the right `nb90_visualvm_xyz.zip` and its `ml` variant by checking out the right sources, building them and adjusting them for purposes of VisualVM.

File: visualvm/build-nb-ml.sh [hunk 2]
```diff
@@ -1,13 +1,76 @@
...
+lib.profiler.common,\
+lib.profiler.ui,\
+profiler,\
+profiler.api,\
+profiler.attach,\
+profiler.heapwalker,\
+profiler.oql,\
+profiler.snaptracer,\
+profiler.utilities"
+PROFILER_LIGHT="$PROFILER_LIGHT -Dvalidation.nb.cluster.profiler=profiler"
```
Comment (JaroslavTulach) [inline visualvm/build-nb-ml.sh]: We redefine the content of `profiler` cluster to skip modules that depend on `java` or `ide` clusters and make the ZIP more lightweight.

File: visualvm/build-nb-ml.sh [hunk 3]
```diff
@@ -1,13 +1,76 @@
...
 ant -Dname=platform rebuild-cluster
 ant -Dname=harness rebuild-cluster
-ant -Dname=profiler rebuild-cluster
-zip -r /tmp/nb90_visualvm_`date "+%d%m%Y"`.zip netbeans
+ant -Dname=profiler rebuild-cluster $PROFILER_LIGHT
+
+# cleanup remote packs and cmd support
+rm -rf netbeans/ide/
+rm -rf netbeans/profiler/remote-pack-defs/
+rm -rf netbeans/profiler/lib/deployed/cmd/
```
Comment (JaroslavTulach) [inline visualvm/build-nb-ml.sh]: Removes the remote packs & etc. support which is not needed by VisualVM.

File: visualvm/build-nb-ml.sh [hunk 4]
```diff
@@ -1,13 +1,76 @@
...
+
+#zip -r $BUILD_ROOT/$ZIPNAME-ml.zip netbeans
 
+rm -rf netbeans
+unzip $BUILD_ROOT/$ZIPNAME.zip
```
Comment (JaroslavTulach) [inline visualvm/build-nb-ml.sh]: Expands the final ZIP into expected location. Thus:
```bash
./build-nb-ml.sh
ant run
```
shall build everything and launch freshly build VisualVM.

File: visualvm/build-nb-ml.sh [hunk 5]
```diff
@@ -1,13 +1,76 @@
-#!/bin/sh
+#!/bin/bash
+
+REV=41da26b7de7506c8
+REV_L10N=1180ea1ceb30
+ZIPNAME=nb90_visualvm_`date "+%d%m%Y"`
+
+set -e
```
Comment (JaroslavTulach) [inline visualvm/build-nb-ml.sh]: If there is an error in the script, fail its execution.

File: README.md [hunk 6]
```diff
@@ -14,7 +14,7 @@ First download or clone this repository into directory `visualvm`. There are two
 
 ## Get the dependencies
   
-Then extract the [NetBeans 9.0 Dev platform and profiler binaries](visualvm/nb90_visualvm_11122017.zip) into directory `visualvm/visualvm` (should create `visualvm/visualvm/netbeans`).
+Then extract the [NetBeans 9.0 Dev platform and profiler binaries](visualvm/nb90_visualvm_01022018.zip) into directory `visualvm/visualvm` (should create `visualvm/visualvm/netbeans`).
```
Comment (JaroslavTulach) [inline README.md]: Replacing the ZIPs with new ones. The sizes seem comparable:
```
28303978 nb90_visualvm_01022018-ml.zip
27318977 nb90_visualvm_01022018.zip
27276414 nb90_visualvm_11122017-ml.zip
26286736 nb90_visualvm_11122017.zip
```
and doing `unzip nb90_visualvm_01022018.zip && ant run` seems to run the app fine.

File: visualvm/build-nb-ml.sh [hunk 7]
```diff
@@ -1,13 +1,76 @@
...
 ant -Dlocales=ja,zh_CN -Dname=platform rebuild-cluster
 ant -Dlocales=ja,zh_CN -Dname=harness rebuild-cluster
-ant -Dlocales=ja,zh_CN -Dname=profiler rebuild-cluster
-zip -r /tmp/nb90_visualvm_`date "+%d%m%Y"`-ml.zip netbeans
+ant -Dlocales=ja,zh_CN -Dname=profiler rebuild-cluster $PROFILER_LIGHT
+
+# cleanup remote packs and cmv support
+rm -rf netbeans/ide/
+rm -rf netbeans/profiler/remote-pack-defs/
+rm -rf netbeans/profiler/lib/deployed/cmv/
```
Comment (thurka) [inline visualvm/build-nb-ml.sh]: Typo `netbeans/profiler/lib/deployed/cmv/` -> 'netbeans/profiler/lib/deployed/cvm/`

File: visualvm/build-nb-ml.sh [hunk 8]
```diff
@@ -1,13 +1,76 @@
...
 ant -Dname=platform rebuild-cluster
 ant -Dname=harness rebuild-cluster
-ant -Dname=profiler rebuild-cluster
-zip -r /tmp/nb90_visualvm_`date "+%d%m%Y"`.zip netbeans
+ant -Dname=profiler rebuild-cluster $PROFILER_LIGHT
+
+# cleanup remote packs and cmv support
+rm -rf netbeans/ide/
+rm -rf netbeans/profiler/remote-pack-defs/
+rm -rf netbeans/profiler/lib/deployed/cmv/
```
Comment (thurka) [inline visualvm/build-nb-ml.sh]: Typo `netbeans/profiler/lib/deployed/cmv/` -> 'netbeans/profiler/lib/deployed/cvm/`

File: README.md (hunk 6 above)
Comment (thurka) [reply README.md]: The size difference looks too big to me. Please investigate what caused this difference.

**Comment by thurka** [comment]
I just built VisualVM on Mac OS X from JaroslavTulach:jtulach/UsingAppearanceController branch and for some reason profiling tab is not displayed.

**Comment by JaroslavTulach** [comment]
Yes, some `/mac` and `/windows` files were missing due to a problem in 1st NetBeans build. Fixed now - by running the `build-nb-ml.sh` script twice - I fix the NetBeans problem tomorrow.

**Comment by JaroslavTulach** [comment]
The problem is in NetBeans build. If one does:
```bash
lib.profiler$ rm -rf *; git checkout -f .; ant nbm
```
then the NBM has just 49 elements - it should have more than one hundred.

**Comment by JaroslavTulach** [comment]
Tomáši, please review [NetBeans $410](https://github.com/apache/incubator-netbeans/pull/410). With that change, it shall be enough to use the existing `build-nb-ml.sh` script to get the right content of the ZIP files.

**Comment by thurka** [comment]
I don't quite understand what was wrong, since the first nb90_visualvm_01022018.zip contained all the missing files and the build from nb90_visualvm_11122017.zip works fine too.

**Comment by JaroslavTulach** [comment]
The files were there, but they were missing from the `update_tracking` list. Thus the `harness`, when building the module, copied just the files listed there and not all of them.

**Comment by thurka** [comment]
Thanks for explanation. I am looking at [bug #410](https://github.com/apache/incubator-netbeans/pull/410) and the lib.profiler/build.xml file looks strange to me. What is external/profiler-external-binaries-8.2.zip? It looks like it is a result of changes done on donation branch, since the old build file worked fine for years and produced correct nbm even for the first run.

**Comment by thurka** [comment]
New NetBeans platform and profiler binaries are significantly bigger than old one. Given the fact that the source files are almost the same, I would like to know, what caused this difference. I asked this question last week, but it disappeared whit the update of the new binaries.

**Comment by JaroslavTulach** [comment]
In the process of donating the code to Apache all binary files were removed - in case of `lib.profiler`, they were put into `external/profiler-external-binaries-8.2.zip`. Whoever did the change however broke the clean generation of the `ant nbm` - it may be better to discuss this issue in [#410](https://github.com/apache/incubator-netbeans/pull/410) where involved people may be watching...

Is the size difference really that big? It is `26M` vs. `26M` as far as I can tell. The content also looks almost the same:
```
$ ls -s nb90_visualvm_11122017.zip -h
26M nb90_visualvm_11122017.zip
$ ls -s nb90_visualvm_05022018.zip -h
26M nb90_visualvm_05022018.zip
$ unzip -v nb90_visualvm_11122017.zip | grep netbeans/ | cut -c59- | sort  >~/a
$ unzip -v nb90_visualvm_05022018.zip  | grep netbeans/ | cut -c59- | sort >~/b
$ diff -u ~/a ~/b
--- /home/devel/a       2018-02-08 17:23:27.845461073 +0100
+++ /home/devel/b       2018-02-08 17:23:36.545499226 +0100
@@ -2,7 +2,6 @@
 netbeans/harness/
 netbeans/harness/antlib/
 netbeans/harness/antlib/bindex-2.2.jar
-netbeans/harness/antlib/jsearch-2.0_05.jar
 netbeans/harness/build.xml
 netbeans/harness/common.xml
 netbeans/harness/config/
@@ -24,7 +23,6 @@
 netbeans/harness/jdk.xml
 netbeans/harness/jnlp/
 netbeans/harness/jnlp/jnlp-launcher.jar
-netbeans/harness/jnlp/jnlp-servlet.jar
 netbeans/harness/jnlp.xml
 netbeans/harness/.lastModified
 netbeans/harness/launchers/
@@ -186,7 +184,6 @@
 netbeans/platform/config/Modules/net-java-html-json.xml
 netbeans/platform/config/Modules/net-java-html-sound.xml
 netbeans/platform/config/Modules/net-java-html.xml
-netbeans/platform/config/Modules/org-jdesktop-layout.xml
 netbeans/platform/config/Modules/org-netbeans-api-annotations-common.xml
 netbeans/platform/config/Modules/org-netbeans-api-htmlui.xml
 netbeans/platform/config/Modules/org-netbeans-api-intent.xml
@@ -281,8 +278,6 @@
 netbeans/platform/docs/
 netbeans/platform/docs/junit-4.12-javadoc.jar
 netbeans/platform/docs/junit-4.12-sources.jar
-netbeans/platform/docs/swing-layout-1.0.4-doc.zip
-netbeans/platform/docs/swing-layout-1.0.4-src.zip
 netbeans/platform/docs/testng-6.8.1-javadoc.zip
 netbeans/platform/.lastModified
 netbeans/platform/lib/
@@ -298,17 +293,14 @@
 netbeans/platform/lib/org-openide-util-ui.jar
 netbeans/platform/modules/
 netbeans/platform/modules/ext/
-netbeans/platform/modules/ext/felix-4.2.1.jar
 netbeans/platform/modules/ext/hamcrest-core-1.3.jar
-netbeans/platform/modules/ext/jhall-2.0_05.jar
 netbeans/platform/modules/ext/jna-platform-4.4.0.jar
 netbeans/platform/modules/ext/jna-4.4.0.jar
 netbeans/platform/modules/ext/junit-4.12.jar
-netbeans/platform/modules/ext/nsProxyAutoConfig.js
+netbeans/platform/modules/ext/org.apache.felix.main-4.2.1.jar
 netbeans/platform/modules/ext/org.eclipse.osgi_3.9.1.v20140110-1610.jar
-netbeans/platform/modules/ext/osgi.cmpn-4.2.jar
+netbeans/platform/modules/ext/org.osgi.compendium-4.2.0.jar
 netbeans/platform/modules/ext/osgi.core-5.0.0.jar
-netbeans/platform/modules/ext/swing-layout-1.0.4.jar
 netbeans/platform/modules/ext/testng-6.8.1-dist.jar
 netbeans/platform/modules/ext/updater.jar
 netbeans/platform/modules/lib/
@@ -329,7 +321,6 @@
 netbeans/platform/modules/net-java-html.jar
 netbeans/platform/modules/net-java-html-json.jar
 netbeans/platform/modules/net-java-html-sound.jar
-netbeans/platform/modules/org-jdesktop-layout.jar
 netbeans/platform/modules/org-netbeans-api-annotations-common.jar
 netbeans/platform/modules/org-netbeans-api-htmlui.jar
 netbeans/platform/modules/org-netbeans-api-intent.jar
@@ -422,7 +413,6 @@
 netbeans/platform/update_tracking/net-java-html-json.xml
 netbeans/platform/update_tracking/net-java-html-sound.xml
 netbeans/platform/update_tracking/net-java-html.xml
-netbeans/platform/update_tracking/org-jdesktop-layout.xml
 netbeans/platform/update_tracking/org-netbeans-api-annotations-common.xml
 netbeans/platform/update_tracking/org-netbeans-api-htmlui.xml
 netbeans/platform/update_tracking/org-netbeans-api-intent.xml
```
The `jdesktop.layout` module was removed as outdated. JavaHelp is currently removed. OSGi packaging changed a bit. Otherwise the ZIPs seem identical.

**Comment by JaroslavTulach** [comment]
Newly generated ZIPs are smaller & the `Timers` class now has the expected size:
```bash
$ ls -s1h nb90_visualvm_08022018*
26M nb90_visualvm_08022018-ml.zip
25M nb90_visualvm_08022018.zip
$ unzip -v netbeans/profiler/lib/jfluid-server-15.jar | grep Timers
    2763  Stored     2763   0% 2018-02-08 18:20 5d712f62  org/netbeans/lib/profiler/server/system/Timers.class
```

**Comment by thurka** [comment]
The size of zip files is OK now.

**Comment by thurka** [comment]
Unfortunately there is missing icon for OQL Console. See attached image.
oql_icon

**Comment by thurka** [comment]
`org/netbeans/modules/profiler/heapwalk/ui/icons/impl/oqlConsole.png` has zero size.

**Comment by JaroslavTulach** [comment]
[#417](https://github.com/apache/incubator-netbeans/pull/417) has been integrated. I am waiting for [#410](https://github.com/apache/incubator-netbeans/pull/410) to be integrated too and then I prepare the new binaries. Thanks for the review.

**Comment by JaroslavTulach** [comment]
New binaries are there. OQL icon seems to be visible. I think I am done.

**Pull request merged**